src/sinks/util/batch.rs

```diff
@@ -1,47 +1,137 @@
+        let max_bytes = match (self.max_bytes, self.max_size) {
+            (Some(_), Some(_)) => return Err(BatchError::BytesAndSize),
+            (Some(bytes), None) => Some(bytes),
+            (None, Some(size)) => Some(size),
+            (None, None) => None,
+        };
+        Ok(Self {
+            max_bytes,
+            max_size: None,
+            ..*self
+        })
+    }
+
+    pub fn use_size_as_events(&self) -> Result<Self, BatchError> {
+        let max_events = match (self.max_events, self.max_size) {
+            (Some(_), Some(_)) => return Err(BatchError::EventsAndSize),
+            (Some(events), None) => Some(events),
+            (None, Some(size)) => Some(size),
+            (None, None) => None,
+        };
+        // Sinks that used `max_size` for an event count cannot count
+        // bytes, so err if `max_bytes` is set.
+        if self.max_bytes.is_some() {
+            return Err(BatchError::BytesNotAllowed);
+        }
+        Ok(Self {
+            max_events: max_events,
+            max_size: None,
+            ..*self
+        })
+    }
+
+    pub fn get_settings_or_default(&self, defaults: BatchSettings) -> BatchSettings {
         BatchSettings {
-            size: self.max_size.unwrap_or(size as usize),
-            timeout: Duration::from_secs(self.timeout_secs.unwrap_or(timeout)),
+            size: BatchSize {
+                bytes: self.max_bytes.unwrap_or(defaults.size.bytes),
+                events: self.max_events.unwrap_or(defaults.size.events),
+            },
+            timeout: self
+                .timeout_secs
+                .map(|secs| Duration::from_secs(secs))
+                .unwrap_or(defaults.timeout),
         }
     }
 }
 
-#[derive(Clone, Debug, Default, Deserialize, Serialize)]
-pub struct BatchEventsConfig {
-    pub max_events: Option<usize>,
-    pub timeout_secs: Option<u64>,
+#[derive(Clone, Copy, Debug, Default)]
+pub struct BatchSize {
+    pub bytes: usize,
+    pub events: usize,
 }
 
-impl BatchEventsConfig {
-    pub fn unwrap_or(&self, size: u64, timeout: u64) -> BatchSettings {
-        BatchSettings {
-            size: self.max_events.unwrap_or(size as usize),
-            timeout: Duration::from_secs(self.timeout_secs.unwrap_or(timeout)),
+#[derive(Clone, Copy, Debug, Default)]
+pub struct BatchSettings {
+    pub size: BatchSize,
+    pub timeout: Duration,
+}
+
+impl BatchSettings {
+    // Fake the builder pattern
+    pub const fn bytes(self, bytes: u64) -> Self {
+        Self {
+            size: BatchSize {
+                bytes: bytes as usize,
+                ..self.size
+            },
+            ..self
+        }
+    }
+    pub const fn events(self, events: usize) -> Self {
+        Self {
+            size: BatchSize {
+                events,
+                ..self.size
+            },
+            ..self
+        }
+    }
+    pub const fn timeout(self, secs: u64) -> Self {
```

Comment:
It seems like this API could be more flexible if it accepted a `Duration`?

Reply:
It could, but at the cost of writing `Duration::from_secs` in all the batch sinks, and nothing uses anything other than seconds.